src/ln/channel.rs

```diff
@@ -2365,40 +2365,40 @@ impl Channel {
 		}
 	}
 
-	pub fn shutdown(&mut self, fee_estimator: &FeeEstimator, msg: &msgs::Shutdown) -> Result<(Option<msgs::Shutdown>, Option<msgs::ClosingSigned>, Vec<(HTLCSource, [u8; 32])>), HandleError> {
+	pub fn shutdown(&mut self, fee_estimator: &FeeEstimator, msg: &msgs::Shutdown) -> Result<(Option<msgs::Shutdown>, Option<msgs::ClosingSigned>, Vec<(HTLCSource, [u8; 32])>), ChannelError> {
 		if self.channel_state & (ChannelState::PeerDisconnected as u32) == ChannelState::PeerDisconnected as u32 {
-			return Err(HandleError{err: "Peer sent shutdown when we needed a channel_reestablish", action: Some(msgs::ErrorAction::SendErrorMessage{msg: msgs::ErrorMessage{data: "Peer sent shutdown when we needed a channel_reestablish".to_string(), channel_id: msg.channel_id}})});
+			return Err(ChannelError::Close("Peer sent shutdown when we needed a channel_reestablish"));
 		}
 		if self.channel_state < ChannelState::FundingSent as u32 {
-			self.channel_state = ChannelState::ShutdownComplete as u32;
-			self.channel_update_count += 1;
-			return Ok((None, None, Vec::new()));
+			// Spec says we should fail the connection, not the channel, but that's nonsense, there
+			// are plenty of reasons you may want to fail a channel pre-funding, and spec says you
+			// can do that via error message without getting a connection fail anyway...
+			return Err(ChannelError::Close("Peer sent shutdown pre-funding generation"));
```

Comment:
"if it hasn't received a funding_signed (if it is a funder) or a funding_created (if it is a fundee): SHOULD fail the connection" if spec is nonsense and it's not mandatory we're free to just close channel maybe ?

Reply:
Yea well the spec is kinda out-of-date, in a few places it references "fail the connection" but thats from originally it assumed one channel per connection, but then they added support for more. So closing the channel should be more than sufficient.

Reply:
Okay